Added ObjectNav task definition, dataset, metrics, goal sensor

mathfac opened the pull request:
## Motivation and Context
Added ObjectNav task definition, dataset, metrics, goal sensor.
- ObjectGoal Nav Task definition: metrics, goals, sensors with documentations.
- ObjectGoal Nav Dataset definition and loading and link for downloading.
- Topdown target objects, view points, bounding boxes visualization
- Compatible with Benchmark API
- Added **ObjectNav DDPPO config for a baseline**
- Made PointNavDataset general enough to be inherited by other Dataset types
- Refactored RLEnv to use measures for reward and success to speed up distance calculations
- Added mechanism for Measures (Metrics) classes to be depend on other measure. That increased speed and calculation and decrease copy pasta. Speed improvements 2x: from 8 FPS to 15.5 FPS, compare to 40 FPS for PointGoal with same setup.
- made SPL dependent on DistanceToGoal
- added metrics calc to reset method for SPL and DistanceToGoal

with @skylion007,  inspired by @erikwijmans's an initial implementation.




<!--- Why is this change required? What problem does it solve? -->
<!--- Please link to an existing issue here if one exists. -->
<!--- (we recommend to have an existing issue for each pull request) -->

## How Has This Been Tested
- Added the unit test.
- [x] Enable the task config for other general tests we have

**Examples of the Object Goal Navigation Episodes**:
Goal semantic category: chair
![ezgif-2-d8bbee870faf](https://user-images.githubusercontent.com/877440/73147597-03590500-406d-11ea-8255-126f6babe211.gif)
Goal semantic category: picture
![ezgif-2-7bc528b0d37a](https://user-images.githubusercontent.com/877440/73147649-4e731800-406d-11ea-8913-217ebe804905.gif)
Goal semantic category: sink
![ezgif-2-23a938625884](https://user-images.githubusercontent.com/877440/73147706-8bd7a580-406d-11ea-9c2f-6d89b16a1d98.gif)
*The target object is highlighted by red overlay on RGB image.*

*Map colors:*
*Red is a center of the target objet.*
*Green is an axes aligned bounding box of the target objet.*
*Red overlay zone near the target objects is view points/success zone.*


<!--- Please describe here how your modifications have been tested. -->

## Types of changes

<!--- What types of changes does your code introduce? Leave all the items that apply: -->
- Docs change / refactoring / dependency upgrade
- New feature (non-breaking change which adds functionality)

erikwijmans commented on configs/tasks/obj_nav_mp3d.yaml:
```diff
@@ -0,0 +1,47 @@
+ENVIRONMENT:
+  MAX_EPISODE_STEPS: 750
```
Why 750?

mathfac replied:
As I remember, to give an option to explore the environment and then choose an object.

erikwijmans commented on configs/tasks/obj_nav_mp3d.yaml:
```diff
@@ -0,0 +1,47 @@
+  MAX_EPISODE_STEPS: 750
+SIMULATOR:
+  TURN_ANGLE: 30
+  TILT_ANGLE: 30
```
We are planning to add look_up and look_down actions to the action space for the challenge?  If so, we should add a max tilt angle parameter.

dhruvbatra replied:
> We are planning to add look_up and look_down actions to the action space for the challenge?

Yes, so the agent can see objects hanging on the wall or placed on the floor.

erikwijmans replied:
Ah, any suggestion for the max title angle then?

mathfac replied:
-90..+90 degrees?

erikwijmans commented on configs/tasks/obj_nav_mp3d.yaml:
```diff
@@ -0,0 +1,47 @@
+  SEMANTIC_SENSOR:
+    WIDTH: 640
+    HEIGHT: 480
+    HFOV: 59
```
Where is the HFOV and width/height coming from?

dhruvbatra replied:
> Where is the HFOV and width/height coming from?

From suggests from AI2 folks (@roozbehm and Ani).

mathfac replied:
Resolution is coming from PyRobot config, HFOV coming from the AI team.

erikwijmans replied:
Ah, they are using a square aspect ratio, so we should set this to 59*4/3 to be consistent (we are using 4:3 and this is hfov).

erikwijmans commented on habitat/tasks/nav/nav.py:
```diff
@@ -453,9 +448,27 @@ def update_metric(
+                current_position, episode.goals[0].position
+            )
+        elif self._config.DISTANCE_TO == "VIEW_POINTS":
+            multi_goal = habitat_sim.MultiGoalShortestPath()
```
This functionality should be pushed into `self._sim.geodesic_distance`

erikwijmans commented on habitat/tasks/nav/nav.py:
```diff
@@ -453,9 +448,27 @@ def update_metric(
+            distance_to_target = multi_goal.geodesic_distance
+
+        # remove after debug
+        # print(
```
Remove

erikwijmans commented on habitat/tasks/nav/nav.py:
```diff
@@ -453,9 +448,27 @@ def update_metric(
+            self._previous_distance_to_target - distance_to_target
+            > self._sim.config.FORWARD_STEP_SIZE
+        ):
+            print("!!! distance_to_target change more than forward step")
```
This is possible with sliding on :) -- so these prints will just be a nuisance.

mathfac replied:
@erikwijmans that was there for debugging purposes. With multi goal setup that was important that the geodesic distance metric is smooth when switching from one closest view point to another.

erikwijmans replied:
Yeah, it'll be smooth.  The algorithm that is doing it in habitat-sim is doing `min(geo_dist(s, view_pt) for view_pt in view_points])`, just in a smart way.

erikwijmans commented on habitat/tasks/nav/object_nav_task.py:
```diff
@@ -0,0 +1,149 @@
+        elif self.config.GOAL_SPEC == "OBJECT_ID":
+            return np.array([episode.goals[0].object_name_id], dtype=np.int64)
+        else:
+            return None
```
None will likely end up causing something else to crash and I don't see why you'd want the object goal sensor to sometimes not return a goal.  These `return None` should likely be `raise RuntimeError(<some error message>)` as that'll make debugging simpler.

erikwijmans commented on habitat/tasks/nav/nav.py:
```diff
@@ -576,18 +589,71 @@ def draw_source_and_target(self, episode):
+            x_len, _, z_len = sem_scene.objects[object_id].aabb.sizes / 2.0
+            corners = [
+                center + np.array([x, 0, z])
+                for x, z in itertools.product([-x_len, x_len], [z_len, -z_len])
```
Perhaps `for x, z in [(-x_len, -z_len), (-x_len, z_len), (x_len, z_len), (x_len, -z_len), (-x_len, -z_len)]` would be better?  That would get rid of `corners[2], corners[3] = corners[3], corners[2]` and `corners.append(corners[0])` (took me a bit to figure out what those where doing).

erikwijmans commented on habitat/tasks/nav/nav.py:
```diff
@@ -440,7 +428,14 @@ def reset_metric(self, *args: Any, episode, **kwargs: Any):
         self._previous_position = self._sim.get_agent_state().position.tolist()
         self._start_end_episode_distance = episode.info["geodesic_distance"]
         self._agent_episode_distance = 0.0
+        self._previous_distance_to_target = 0.0  # remove after debug
```
^

erikwijmans commented on habitat/datasets/object_nav/object_nav_dataset.py:
```diff
@@ -0,0 +1,97 @@
+                episode.scene_id = os.path.join(scenes_dir, episode.scene_id)
+
+            for i in range(len(episode.goals)):
+                if "best_iou" in episode.goals[i]:  # remove before release
```
Is this line suppose to stay?

mathfac replied:
Legacy should be removed till we really need iou.

smorad commented:
This is relevant to my interests, is there an associated dataset for this?

dhruvbatra commented:
Dataset and challenge details for CVPR20 challenge/workshop are coming soon.

mathfac commented:
Address the comments, refactored drawing functions, update the config. PR depends on https://github.com/facebookresearch/habitat-api/pull/290.

smorad commented:
Does "compatible with benchmark api" refer to `habitat_baselines`? If so, could you include the baseline `.yaml` you created for testing purposes?

abhiskk commented on habitat/datasets/object_nav/object_nav_dataset.py:
```diff
@@ -0,0 +1,95 @@
+        )
+
+        assert set(self.category_to_task_category_id.keys()) == set(
+            self.category_to_mp3d_category_id.keys()
```
Should we have Matterport3d (`mp3d`) specific variable names here, is this implementation dataset (Matterport3D, Gibson) agnostic?

mathfac replied:
That's good question. We can change `category_to_mp3d_category_id` to `category_to_scene_annotation_category_id` and then use scene dataset type for further category conversions. cc @Skylion007

abhiskk commented on habitat/tasks/nav/nav.py:
```diff
@@ -614,7 +664,7 @@ def reset_metric(self, *args: Any, episode, **kwargs: Any):
                     self._coordinate_min,
                     self._coordinate_max,
                     self._map_resolution,
-                )[::-1]
```
What is the reason for getting rid of the list reversal here?

mathfac replied:
Because, `maps.draw_path` method was changed that switching x and y isn't needed anymore. That makes `maps.draw_path` easier to use.

abhiskk commented on habitat/utils/visualizations/maps.py:
```diff
@@ -47,6 +49,8 @@
 TOP_DOWN_MAP_COLORS[MAP_SOURCE_POINT_INDICATOR] = [0, 0, 200]
 TOP_DOWN_MAP_COLORS[MAP_TARGET_POINT_INDICATOR] = [200, 0, 0]
 TOP_DOWN_MAP_COLORS[MAP_SHORTEST_PATH_COLOR] = [0, 200, 0]
+TOP_DOWN_MAP_COLORS[MAP_VIEW_POINT_INDICATOR] = [245, 150, 150]
```
nit: if possible add color names here, it would be useful for someone reading the codebase.

mathfac commented on configs/tasks/objectnav_mp3d.yaml:
```diff
@@ -0,0 +1,47 @@
+    WIDTH: 640
+    HEIGHT: 480
+    HFOV: 79
+    MIN_DEPTH: 0.5
```
Yes, that is motivated by real depth camera.

mathfac commented on habitat/tasks/nav/nav.py:
```diff
@@ -559,51 +565,95 @@ def get_original_map(self):
 
         return top_down_map
 
-    def draw_source_and_target(self, episode):
```
It was broke into `_draw_point` and `_draw_goals_positions` as we start drawing more sophisticated goals.

mathfac commented:
- Made PointNavDataset general enough to be inherited
- Refactored RLEnv to use measures for reward and success to speed up distance calculations
- made SPL dependent on DistanceToGoal
- added metrics calc to reset method for SPL and DistanceToGoal

mathfac commented:
- Added doc to FloatJSON encoder, 
- Added color names
- Removed mp3d mentioning from ObjectNavDataset class

erikwijmans commented on habitat_baselines/config/objectnav/ddppo_objectnav.yaml:
```diff
@@ -0,0 +1,60 @@
+VIDEO_OPTION: ["disk", "tensorboard"]
+TENSORBOARD_DIR: "/checkpoint/maksymets/logs/habitat_baselines/ddppo/obj_nav_mp3d"
+VIDEO_DIR: "data/video_dir"
+TEST_EPISODE_COUNT: 2
```
Let's put the actual number of val episodes here.  It is hard to figure out what number should go here when you want to evaluate on everything otherwise.
